test/gleam/time/timestamp_test.gleam

```diff
@@ -207,3 +214,329 @@ pub fn to_rfc3339_8_test() {
+// the caller must ensure good values are passed in.
+fn parse_rfc3339_oracle(input: String) -> Result(timestamp.Timestamp, Nil) {
+  use total_milliseconds <- result.map(rfc3339_to_system_time_in_milliseconds(
+    input,
+  ))
+  // Break out the millisecond fraction first so that the conversion to
+  // nanoseconds doesn't overflow for JS in the normalise function.
+  let millisecond_fraction = total_milliseconds % 1000
+  let whole_seconds = { total_milliseconds - millisecond_fraction } / 1000
+  timestamp.from_unix_seconds_and_nanoseconds(
+    seconds: whole_seconds,
+    nanoseconds: millisecond_fraction * 1_000_000,
+  )
+}
+
+pub fn parse_rfc3339_matches_oracle_example_0_test() {
+  parse_rfc3339_should_match_oracle_for("9999-12-31T23:59:59.999Z")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_1_test() {
+  parse_rfc3339_should_match_oracle_for("1970-01-01T00:00:00.111Z")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_2_test() {
+  parse_rfc3339_should_match_oracle_for("1970-01-01T00:00:00.000Z")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_3_test() {
+  parse_rfc3339_should_match_oracle_for("1969-12-31T23:59:59.444Z")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_4_test() {
+  parse_rfc3339_should_match_oracle_for("1969-12-31T23:59:58.666Z")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_5_test() {
+  parse_rfc3339_should_match_oracle_for("0000-01-01T00:00:00Z")
+}
+
+@target(javascript)
+pub fn parse_rfc3339_matches_oracle_example_6_test() {
+  // The oracle gives badarg on Erlang as it is beyond the range of the
+  // 0000-01-01T00:00:00Z limit.
+  parse_rfc3339_should_match_oracle_for("0000-01-01T00:00:00+00:01")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_7_test() {
+  parse_rfc3339_should_match_oracle_for("0000-01-01T00:00:00-00:01")
+}
+
+pub fn parse_rfc3339_matches_oracle_example_8_test() {
+  parse_rfc3339_should_match_oracle_for("9999-12-31T23:59:59.999+00:01")
+}
+
+@target(javascript)
+pub fn parse_rfc3339_matches_oracle_example_9_test() {
+  // This oracle gives badarg on Erlang as it is beyond the range of the
+  // 9999-12-31T23:59:59.999Z limit.
+  parse_rfc3339_should_match_oracle_for("9999-12-31T23:59:59.999-00:01")
+}
+
+@target(erlang)
+pub fn parse_rfc3339_matches_oracle_example_10_test() {
+  // JS returns NaN for any leap seconds, so skip this test in JS.
+  parse_rfc3339_should_match_oracle_for("1970-01-01T23:59:60Z")
+}
+
+fn parse_rfc3339_should_match_oracle_for(date_time) {
+  should.equal(
+    timestamp.parse_rfc3339(date_time),
+    parse_rfc3339_oracle(date_time),
+  )
+}
+
+pub fn parse_rfc3339_matches_oracle_property_test() {
+  use date_time <- qcheck.given(rfc3339_generator.date_time_generator(
+    with_leap_second: False,
+    secfrac_spec: rfc3339_generator.WithMaxLength(3),
+  ))
+
+  let result = timestamp.parse_rfc3339(date_time)
+  let expected = parse_rfc3339_oracle(date_time)
+  case result, expected {
```

Comment:
This test is hard too complex and hard to read. Can we filter these out beforehand in the generator please

Reply:
Yes good point--I think not generating them in the first place, and then using a couple of well chosen example tests will be better here.